spec/index.adoc

```diff
@@ -1637,11 +1637,13 @@ For security reasons, only HTTPS connections are allowed (URLs must start with `
 
 The maximal size of a request URL is 2048 bytes.
 
-The maximal size of a response is `2MiB`. Therefore, only the first `2MiB` will be returned if a response is larger than this size. This size limit also applies to the value returned by the `transform` function.
+Each request should specify the maximal expected size for the response. The maximal size of a response is `2MiB`. Therefore, only the first number of bytes as specified in the request (up to 2MiB), will be returned if a response is larger than these limits. The 2MiB size limit also applies to the value returned by the `transform` function.
```

Comment:
Silent truncation sounds like a footgun.

Can’t we return an error in that case? HTTP error 413 isn’t quite the right thing, because that seems to be about the request size. Maybe 206 Partial Content?

Or maybe better to issue an IC-level reject here.

Reply:
This is not that silent if the user specifies the maximal size.
The idea is to encourage users to send requests that result in smaller responses, as we are worried that this feature could use lots of bandwidth otherwise. I do see cases where the first X bytes of a bigger response are important.
We do not want to reject the call if the first X requested bytes can be served..

Reply:
But it is silent! How do I know whether the request was truncated, or just happened to have the size I was expecting? And if the user does not specify the size explicitly, it’s still 2MiB, right?

>  We do not want to reject the call if the first X requested bytes can be served..

Would that use-case be better serves using the `Ranges:` header of HTTP? This is an existing protocol lettings users request parts of a response?

We could either hope that most servers support this, or implement that ranges in the HTTP component.

Or at least add an `IC-response-truncated:` header maybe? Still a footgun, I’d say.

Reply:
As an engineer, I see little usage in receiving only part of a response, tbh. Thus, I agree that we should fail hard—and early.

Reply:
Users could use ranges in their request.
I meant that it is not silent as the user specifies the size and the response should have a `Content-Length` header so they can know if it was truncated. But I also see the un-usefulness in this kind of truncated response...

Reply:
Is that header guaranteed to be there? If the server streams responses probably not.

Reply:
Indeed, if `Transfer-Encoding` is set, `Content-Length` must not be set ...

https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html

While for most common use cases, Content-Length is set—there are sloppy http implementations out there.

Reply:
I agree with @nomeata's concern. We should not discard silently.

Reply:
I also think it's better to reject if the response does not fit the limit provided. Having this parameter that the users can set makes them more aware of the limitations.

Reply:
Please see updated text